Add TS configuration and NX project for metadata

IlyaKhD opened the pull request:
## Configuration Helpers for Mutations

```ts
[
  removeMembers(/\/card_view:Toolbar/)
]
```

```ts
[
  addMember({
    uid: 'ui/popover:dxPopoverOptions.toolbarItems',
    types: [array(memberRef('ui/popover:ToolbarItem'))],
  }),
]
```

```ts
[
  ...replaceWithWidgetFactory({
    uid: 'ui/form:dxFormSimpleItem',
    from: {
      componentNameProp: 'editorType',
      componentConfigProp: 'editorOptions',
    },
    to: {
      factoryName: 'FormItemEditor',
      newProp: 'editor',
    }
  })
]
```

IlyaKhD committed:
Add NX project for metadata

pomahtri commented on packages/metadata/project.json:
```diff
@@ -0,0 +1,97 @@
+        "{workspaceRoot}/packages/devextreme/js/**/*.js"
+      ],
+      "outputs": [
+        "{projectRoot}/artifacts/Declarations.json",
```
Let's set whole `artifacts` folder as output, so that we will not forget to update this config

IlyaKhD replied:
Hmmm, but multiple tasks output to this folder, I feel like it gonna be a mess with caching

IlyaKhD replied:
Maybe I misunderstood the idea. Here’s how I see specifying the common output folder:

<img width="839" alt="image" src="https://github.com/user-attachments/assets/5bf4bf2c-700c-4573-9e8c-80e5bd00a5a2" />

https://nx.dev/recipes/running-tasks/configure-outputs#workspace-level-outputs

IlyaKhD replied:
But I agree there's a maintenance risk of forgetting to update the artifacts list
It's too bad that NX doesn't offer an option for TS project configurations

pomahtri commented on packages/metadata/project.json:
```diff
@@ -0,0 +1,97 @@
+      "cache": false,
+      "inputs": [
+        "metadataToolsCommonInputs",
+        "{workspaceRoot}/packages/devextreme/js/**/*.ts",
```
Please check that this cache works good, as far as I remember, NX can work not good if inputs of different projects have intersection
Also please consider adding whole devextreme into `implicitDependencies`

IlyaKhD replied:
Well, I actually planned to disable caching for now, since it has to carefully checked.
Still, I specified the inputs and outputs to give some initial context to whoever handles caching in the future

pomahtri commented on packages/devextreme-vue/project.json:
```diff
@@ -22,7 +22,7 @@
         "script": "regenerate"
       },
-      "dependsOn": ["devextreme-monorepo:\"vue:copy-metadata\""]
+      "dependsOn": ["metadata:\"put-metadata-to-vue\""]
```
Let's not add tasks like `put-smth-to-smth`, instead  let's add coping or importing (importing is better) into build/generate task

IlyaKhD replied:
Well, this PR was meant to convert metadata configurations to TS, so I wanted to make it in a simple way.

But I agree - if a package created, it should be set up properly

IlyaKhD replied:
Changed

pomahtri commented:
You may want to check my old draft https://github.com/pomahtri/DevExtreme/compare/fae63049bad2b896d4b45c983e05bc1358e8d45a...refactor/nx_projects/metadata

IlyaKhD committed:
refactor(metadata): convert to Nx library package and clean up build process

- Change projectType from 'application' to 'library' in project.json
- Remove unused outputs (devextreme-uml.wsd, ts-members-meta.json) from discover-declarations
- Ensure build and pack tasks use correct inputs/outputs and glob patterns
- Update build script to only copy required artifacts
- General cleanup for Nx and npm package best practices

IlyaKhD committed:
refactor(metadata): replace copying of JSON artifacts with direct require-based references

- Add "exports" field to metadata/package.json to explicitly expose JSON files from dist/
- Update Angular, React, and Vue consumers to use require.resolve('metadata/…') for JSON files instead of copying them
- Simplifies and unifies metadata consumption across the monorepo

IlyaKhD committed:
refactor(metadata): remove build script, rename artifacts to dist, update references

- Remove obsolete build script; metadata now outputs directly to dist
- Rename artifacts dir to dist and update all related configs and ignores

IlyaKhD committed:
refactor: rename metadata package to devextreme-metadata

IlyaKhD committed:
fix: fix devextreme-metadata devDependency in all consumers

IlyaKhD committed:
chore: rename metadata folder to devextreme-metadata

IlyaKhD committed:
fix: update inject-descriptions scripts to use devextreme-metadata

IlyaKhD committed:
refactoring: update remaining workflows and scripts for devextreme-metadata artifact migration

IlyaKhD committed:
fix: update the build-all script to use devextreme-metadata project name

IlyaKhD committed:
Merge remote-tracking branch 'dx/25_1' into khd/generation-ts-config

# Conflicts:
#	package.json
#	tools/imd-cfg.json
#	tools/smd-cfg.json

IlyaKhD committed:
Update metadata configurations after pull

IlyaKhD committed:
Refactor mutation helpers

IlyaKhD committed:
Bump devextreme-internal-tools

IlyaKhD committed:
refactor: redirect references to devextreme-metadata

IlyaKhD committed:
Clean up the project.json

IlyaKhD committed:
Fix lint

IlyaKhD committed:
Use inline require for devextreme-metata paths

IlyaKhD commented:
> You may want to check my old draft [pomahtri/DevExtreme@fae6304...refactor/nx_projects/metadata](https://github.com/pomahtri/DevExtreme/compare/fae63049bad2b896d4b45c983e05bc1358e8d45a...refactor/nx_projects/metadata)

Yes, I was just about to ask for it, thnx 🤝

The aim of this PR is to use TypeScript metadata configurations, but I agree we should make it more aligned with the NX way, since we have the opportunity

IlyaKhD committed:
Replace path-related constants with functions

IlyaKhD committed:
Remove centralization for devextreme-metadata paths
